state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1107,35 +1231,228 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		Ok(())
 	}
 
+	/// Returns participating pools and the total boost fee if successful
+	#[transactional]
+	fn try_boosting(
+		asset: TargetChainAsset<T, I>,
+		required_amount: TargetChainAmount<T, I>,
+		max_boost_fee: BasisPoints,
+		boost_id: u64,
+	) -> Result<(Vec<BoostPoolTier>, TargetChainAmount<T, I>), DispatchError> {
+		let mut remaining_amount = required_amount;
+
+		let mut total_fee_amount: TargetChainAmount<T, I> = 0u32.into();
+
+		let mut used_pools = vec![];
+
+		for boost_tier in SORTED_BOOST_TIERS {
+			if boost_tier as u16 > max_boost_fee {
+				break
+			}
+
+			// For each fee tier, get the amount that the pool is boosting and the boost fee
+			let (provided_amount, fee) = BoostPools::<T, I>::mutate(asset, boost_tier, |pool| {
+				let pool = match pool {
+					Some(pool)
+						if pool.get_available_amount() == TargetChainAmount::<T, I>::from(0u32) =>
+					{
+						return Ok::<_, DispatchError>((0u32.into(), 0u32.into()));
+					},
+					Some(pool) => pool,
+					None => {
+						// Pool not existing for some reason is equivalent to not having funds:
+						return Ok::<_, DispatchError>((0u32.into(), 0u32.into()));
+					},
+				};
+
+				let full_amount_fee = Self::compute_fee(remaining_amount, boost_tier as u16);
+
+				// How much is needed after pool's fee is deducted:
+				let required_amount = remaining_amount.saturating_sub(full_amount_fee);
+
+				used_pools.push(boost_tier);
+
+				if pool.get_available_amount() >= required_amount.saturating_sub(full_amount_fee) {
```

Comment:
Just realised that `full_amount_fee` is "deducted" twice when checking here. I'm updating this code though (moving the fee calculation inside of `use_funds_for_boosting` and that should address this as well.